idl/chromadb/proto/chroma.proto

```diff
@@ -87,19 +87,141 @@ message OperationRecord {
+/* Metadata Reader Interface */
+
+service MetadataReader {
+    rpc GetMetadata(GetMetadataRequest) returns (GetMetadataResponse) {}
 }
 
-message VectorQueryResult {
+message GetMetadataRequest {
+    Where where = 1;
+    WhereDocument where_document = 2;
+    repeated string ids = 3;
+    optional int32 limit = 4;
+    optional int32 offset = 5;
+}
+
+message GetMetadataResponse {
+    repeated MetadataEmbeddingRecord records = 1;
+}
+
+message MetadataEmbeddingRecord {
     string id = 1;
-    float distance = 3;
-    optional Vector vector = 4;
+    UpdateMetadata metadata = 2;
 }
 
-message VectorQueryResults {
-    repeated VectorQueryResult results = 1;
+// A `WhereDocument` clause for filtering embeddings based on their associated documents.
+message WhereDocument {
+    repeated string contains = 1;
+    repeated string not_contains = 2;
+}
+
+// A `Where` clause for filtering metadata. A `Where` clause is a tree of
+// `Where` clauses, where each node is exactly one of:
+// - A leaf node representing a direct comparison between a metadata key and a
+//   value or list of values.
+// - An branch node with a list of children and a way to combine them (AND or OR).
+message Where {
+    oneof where {
+        DirectComparison direct_comparison = 1;
+        WhereChildren children = 2;
+    }
+}
+
+message DirectComparison {
+    oneof comparison {
+        SingleStringComparison single_string_operand = 1;
+        StringListComparison string_list_operand = 2;
+        SingleIntComparison single_int_operand = 3;
+        IntListComparison int_list_operand = 4;
+        SingleFloatComparison single_float_operand = 5;
+        FloatListComparison float_list_operand = 6;
+    }
+}
+
+message WhereChildren {
+    repeated Where children = 1;
+    WhereChildrenType children_type = 2;
+}
+
+// A `Where` clause may have a list of children. This enum specifies how the
+// children should be combined.
+enum WhereChildrenType {
+    AND = 0;
+    OR = 1;
+}
+
+// A `Where` clause may have a list of allowed or disallowed values. This enum
+// specifies which type of list it is.
+enum WhereListType {
+    IN = 0;
+    NIN = 1;
+}
+
+// A leaf-node `Where` clause may compare a string, int, or float to a single
+// value of the same type. These comparators apply to all three of those types.
+enum GenericComparator {
+    EQ = 0;
+    NE = 1;
+}
+
+// Used when a leaf-node `Where` clause compares an int or float to a single
+// value of the same type.
+enum NumberComparator {
+    GT = 0;
+    GTE = 1;
+    LT = 2;
+    LTE = 3;
+}
+
+// Used when a leaf-node `Where` clause compares a string to a list of strings.
+// `WhereListType` specifies whether values in the list are allowed or disallowed.
+message StringListComparison {
+    string key = 1;
+    repeated string values = 2;
+    WhereListType list_type = 3;
+}
+
+// Used when a leaf-node `Where` clause compares a string to a single string.
+message SingleStringComparison {
```

Comment:
thinking: for transport and simplicity do we want to just make the single case a list of n=1?

Reply:
I'm fine either way. I don't see either one as particularly more simple or complex -- both are a little ugly lmao. Which would you prefer?